### PR TITLE
Add TaggedMetricsServiceInvocationEventHandler

### PR DESCRIPTION
**Pull request by mglazer**

### CODE REVIEW

**Review by schlosna — APPROVED** [review]

File: tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+                .update(nanos, TimeUnit.NANOSECONDS);
+    }
+
+    @Override
+    public final void onFailure(@Nullable InvocationContext context, @Nonnull Throwable cause) {
+        markGlobalFailure();
+        if (context == null) {
+            logger.debug(
+                    "Encountered null metric context likely due to exception in preInvocation: {}",
+                    UnsafeArg.of("exception", cause),
```
Comment (carterkozak) [inline tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java]: this is unnecessary, we're providing the full Throwable in the next arg

File: tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    public TaggedMetricsServiceInvocationEventHandler(
+            TaggedMetricRegistry taggedMetricRegistry,
+            String serviceName) {
+        super(getEnabledSupplier(serviceName));
+        this.taggedMetricRegistry = checkNotNull(taggedMetricRegistry, "metricRegistry");
+        this.serviceName = checkNotNull(serviceName, "serviceName");
+    }
+
+    private static String failuresMetricName() {
+        return "failures";
```
Comment (carterkozak) [inline tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java]: any reason not to define a constant instead?

**Comment by carterkozak** [comment]
Should we add these apis?

`Instrumentation.withMetrics(TaggedMetricRegistry metricRegistry)`
`Tritium.instrument(Class<T> serviceInterface, U delegate, TaggedMetricRegistry metricRegistry)`

File: tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java (hunk 1 above)
Comment (mglazer) [reply tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java]: I'll make the fix to the repo where I copied this....

File: tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java (hunk 2 above)
Comment (mglazer) [reply tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java]: Done

File: tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                .safeName(serviceName + "-" + FAILURES_METRIC_NAME)
+                .putSafeTags("service-name", context.getMethod().getDeclaringClass().getSimpleName())
+                .putSafeTags("endpoint", context.getMethod().getName())
+                .putSafeTags("cause", cause.getClass().getName())
+                .build();
+        taggedMetricRegistry.meter(failuresMetricName).mark();
+    }
+
+    private void markGlobalFailure() {
+        taggedMetricRegistry.meter(MetricName.builder().safeName(FAILURES_METRIC_NAME).build()).mark();
```
Comment (carterkozak) [inline tritium-metrics/src/main/java/com/palantir/tritium/event/metrics/TaggedMetricsServiceInvocationEventHandler.java]: nit: the MetricName can be a static constant

**Review by carterkozak — APPROVED** [review]
lgtm when the linter is happy

**Review by schlosna — APPROVED** [review]

File: tritium-lib/src/main/java/com/palantir/tritium/proxy/Instrumentation.java [hunk 4]
```diff
@@ -134,6 +136,27 @@ private Builder(Class<T> interfaceClass, U delegate) {
...
+         */
+        public Builder<T, U> withTaggedMetrics(TaggedMetricRegistry metricRegistry, String globalPrefix) {
+            checkNotNull(metricRegistry, "metricRegistry");
+            this.handlers.add(new TaggedMetricsServiceInvocationEventHandler(
+                    metricRegistry, globalPrefix));
+            return this;
+        }
+
+        public Builder<T, U> withTaggedMetrics(TaggedMetricRegistry metricRegistry) {
+            return withTaggedMetrics(metricRegistry, null);
```
Comment (carterkozak) [inline tritium-lib/src/main/java/com/palantir/tritium/proxy/Instrumentation.java:157]: Does this always NPE?

Comment (schlosna) [reply tritium-lib/src/main/java/com/palantir/tritium/proxy/Instrumentation.java:157]: yes 👎, see https://github.com/palantir/tritium/pull/153 to fix

**Pull request merged**